src/providers/solana/cachedRpcFactory.ts

```diff
@@ -0,0 +1,126 @@
+    const superParams = rateLimitedConstructorParams.slice(-2) as [
+      ConstructorParameters<typeof SolanaClusterRpcFactory>[0], // clusterUrl: ClusterUrl
+      ConstructorParameters<typeof SolanaClusterRpcFactory>[1], // chainId: number
+    ];
+    super(...superParams);
+
+    // Create the rate limited transport and RPC client.
+    const rateLimitedRpcFactory = new RateLimitedSolanaRpcFactory(...rateLimitedConstructorParams);
+    this.rateLimitedTransport = rateLimitedRpcFactory.createTransport();
+    this.rateLimitedRpcClient = rateLimitedRpcFactory.createRpcClient();
+
+    // Pre-compute as much of the redis key as possible.
+    const cachePrefix = `${providerCacheNamespace},${new URL(this.clusterUrl).hostname},${this.chainId}`;
+    this.getTransactionCachePrefix = `${cachePrefix}:getTransaction,`;
+  }
+
+  public createTransport(): RpcTransport {
+    return async <TResponse>(...args: Parameters<RpcTransport>): Promise<TResponse> => {
+      const { method, params } = args[0].payload as { method: string; params?: unknown[] };
+      const cacheType = this.redisClient ? this.cacheType(method) : CacheType.NONE;
+
+      if (cacheType !== CacheType.NONE) {
+        const redisKey = this.buildRedisKey(method, params);
+
+        // Attempt to pull the result from the cache.
+        const redisResult = await this.redisClient?.get<string>(redisKey);
+
+        // If cache has the result, parse the json and return it.
+        if (redisResult) {
+          return JSON.parse(redisResult, jsonReviverWithBigInts);
+        }
+
+        // Cache does not have the result. Query it directly and cache it if finalized.
+        return this.requestAndCacheFinalized<TResponse>(...args);
+      }
+
+      return this.rateLimitedTransport<TResponse>(...args);
+    };
+  }
+
+  private async requestAndCacheFinalized<TResponse>(...args: Parameters<RpcTransport>): Promise<TResponse> {
+    const { method, params } = args[0].payload as { method: string; params?: unknown[] };
+
+    // Only handles getTransaction right now.
+    if (method === "getTransaction") {
+      // Do not throw if params are not valid, just skip caching and pass through to the underlying transport.
+      if (!this.isGetTransactionParams(params)) return this.rateLimitedTransport<TResponse>(...args);
+
+      // Check the confirmation status first to avoid caching non-finalized transactions.
+      const getSignatureStatusesResponse = await this.rateLimitedRpcClient
+        .getSignatureStatuses([params[0]], {
+          searchTransactionHistory: true,
+        })
+        .send();
+
+      const getTransactionResponse = await this.rateLimitedTransport<TResponse>(...args);
+
+      // Cache the transaction only if it is finalized.
+      if (getSignatureStatusesResponse.value[0]?.confirmationStatus === "finalized") {
+        const redisKey = this.buildRedisKey(method, params);
+        await this.redisClient?.set(
```

Comment:
I think there is no need to await here since it won't affect the returned value

Reply:
agree, but that would make error handling harder. And we also await in the EVM implementation, so maybe can keep it here for consistency?